packages/@lwc/engine-dom/src/apis/hydrate-component.ts

```diff
@@ -5,10 +5,43 @@
  * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
  */
 
-import { createVM, LightningElement, hydrateRootElement } from '@lwc/engine-core';
+import {
+    createVM,
+    LightningElement,
+    hydrateRootElement,
+    connectRootElement,
+    getAssociatedVMIfPresent,
+} from '@lwc/engine-core';
 import { isFunction, isNull, isObject } from '@lwc/shared';
 import { setIsHydrating } from '../renderer';
-import { createElement } from './create-element';
+
+function resetShadowRootAndLightDom(element: Element, Ctor: typeof LightningElement) {
+    if (element.shadowRoot) {
+        const shadowRoot = element.shadowRoot;
+
+        while (!isNull(shadowRoot.firstChild)) {
+            shadowRoot.removeChild(shadowRoot.firstChild);
+        }
```

Comment:
I don't know the perf implications, but this is shorter:

```suggestion
        shadowRoot.innerHTML = '';
```

Reply:
Remove child is more performant than innerHTML: https://www.measurethat.net/Benchmarks/Show/34/0/innerhtml-vs-removechild